### PR TITLE
rsmeter: distinguish between collectors and observers within recorder 

### LINKED ISSUE #11689: resource-metering: improve the integration test of scanned-keys
## Development Task

This PR (https://github.com/tikv/tikv/pull/11085) introduces the ScanRow and provides some integration tests. But the test does not work well on different machines. The reason is that the fixed sleep wait may not get results on slower machines. This issue is used to track the optimization of the integration test for this PR.

### PR DESCRIPTION
**Pull request by zhongzc**
### What is changed and how it works?

ref #11691

close #11689 

What's Changed:

* Add `as_observer` as an argument to register `Collector`
* Recorder as a runnable worker
* Recorder runs only if existing `Collector`

### Check List 

Tests 

- Unit test
- Integration test

### Release note 

```release-note
none
```

### CODE REVIEW

File: components/resource_metering/src/recorder/sub_recorder/mod.rs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+    fn pause(
+        &mut self,
+        _records: &mut RawRecords,
+        _thread_stores: &mut HashMap<usize, LocalStorage>,
+    ) {
+    }
+
+    /// This function is called when a reset is required.
+    ///
+    /// The typical situation is when the [Recorder] thread is resume execution.
```
Comment (mornyx) [inline components/resource_metering/src/recorder/sub_recorder/mod.rs:68]:
```suggestion
    /// This function is called when the [Recorder] thread is resume execution.
```

File: components/resource_metering/src/recorder/mod.rs [hunk 2]
```diff
@@ -1,102 +1,66 @@
...
-    /// [LocalStorage]: crate::localstorage::LocalStorage
-    fn collect(
-        &mut self,
-        _records: &mut RawRecords,
-        _thread_stores: &mut HashMap<usize, LocalStorage>,
-    ) {
+    fn run(&mut self, task: Self::Task) {
+        match task {
+            Task::CollectorReg(reg) => self.handle_collector_registration(reg),
+            Task::NewThread(lsr) => self.handle_thread_registration(lsr),
```
Comment (mornyx) [inline components/resource_metering/src/recorder/mod.rs:39]: How about keeping them name consistent, such as `CollectorReg/ThreadReg` or `NewCollector/NewThread`.

**Review by mornyx — COMMENTED** [review]
LGTM

File: components/resource_metering/src/config.rs [hunk 3]
```diff
@@ -20,9 +20,12 @@ const MIN_REPORT_RECEIVER_INTERVAL: ReadableDuration = ReadableDuration::millis(
...
 #[serde(rename_all = "kebab-case")]
 pub struct Config {
-    /// Turn on resource metering.
+    /// `enabled` is deprecated.
     ///
-    /// This configuration will affect all resource modules such as cpu/summary.
+    /// Whether resource metering is enabled depends on the number of non-observed
+    /// collectors. See [CollectorRegHandle::register].
+    ///
+    /// [CollectorRegHandle::register]: crate::CollectorRegHandle::register
```
Comment (breeswish) [inline components/resource_metering/src/config.rs:28]: Can we drop it?

**Review by breeswish — APPROVED** [review]

**Review by andylokandy — APPROVED** [review]

**Pull request merged**